Consistent Autocorrelation and Intermittency - Across Different Analysis

bieniekmateusz opened the pull request:
Fixes #2247 

Changes made in this Pull Request:
 - Provides well defined autocorrelation which can be used for both, SurvivalProbability and HydrogenLifetimes (and more?)
 - Autocorrelation function and intermittency function are now independant with their own test cases
 - Removed the class HydrogenBondLifetimes and moved the autocorrelation functionality into the hbond_analysis class as a .autocorrelation  method. This method should replace also hbond_autocorrel.py file in the future. 
 - New test cases for the method .autocorrelation (HydrogenBondAnalysis) were added 
 - Added a test cases that, with our definition of intermittency, would fail in the case of the hbond_autocorrel.py.

The main issue with HydrogenBondLifetimes (besides the bug) is that it defined its own definition of a hydrogen bond, without allowing the user to change it. It should be clear that the hydrogen bonds first have to be computed by the user, and then the autocorrelation for the bonds can be found. This is achieved with the method .autocorrelation. 

PR Checklist
------------
 - [X] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [X] Issue raised/referenced?

Work done in duo @p-j-smith

bieniekmateusz commented:
It might be good to consider joining this later with PR #2237 by @p-j-smith, which should be easy to do since the functions stand alone. 

Linking #2238 for discussion.

bieniekmateusz commented on package/MDAnalysis/analysis/hbonds/hbond_autocorrel.py:
```diff
@@ -296,6 +296,11 @@ def __init__(self, universe,
                  nsamples=50,  # number of different points to sample in a run
                  pbc=True):
+
+        warnings.warn("This class is deprecated, use analysis.hbonds.HydrogenBondAnalysis "
```
We can merge the changes together from #2237 and here and submit it as one large commit. The two converge together but we tried to separate them and make them incremental to simplify the decision making. 

The definition of hbonds in HBL was defined internally (hiding it from the user) and could not be changed previously from the point of view of the user, which I think is problematic and could lead to a naive use of the class. This, and due to the existing bug #2247, and the overall design of the code, where only one Hydrogen Bond analysis tool exists (analysis.hbonds.HydrogenBondAnalysis) would be much cleaner. 

Could I maybe tell the user to run analysis.hbonds.HydrogenBondAnalysis and explicitly give the user the parameters that were previously used in HBL? This way they could reproduce partially the same analysis.

richardjgowers replied:
@bieniekmateusz I'd rather do the opposite, get #2237 done (which is large enough as is), then rebase these changes on top of that and see what these changes then look like.

One other concern I've got is performance, which is easier to make decisions on once hbondanalysis is given a fair chance of being fast.

bieniekmateusz replied:
I'd hope the changes to be minimal. We have a clear interface for autocorrelation and would simply plug that into the new hbonds.HydrogenBondAnalysis. That's partly why we're doing this - to simplify it. We would have to replace only one line in the new code. 

Performance wise, we're not making anything slower. Previously, HBL used internally hbonds.HydrogenBondAnalysis, and the code for autocorrelation I am certain was slower (the overhead of function calling, and multiple pass over the dataset, in comparison to our linear implementation with the "window").

p-j-smith replied:
Sorry for the confusion over the deprecation warning - we understood that we should first implement the hydrogen bond lifetime functionality in the new hydrogen bond class before pressing ahead with this PR. The avoid over-complicating this current pull request, we've removed this deprecation warning and will add a hydrogen bond lifetime method that uses this autocorrelation function in a future pull request.

bieniekmateusz replied:
Hey @richardjgowers, we temporarily removed the changed made to the part of the code and the deprecation warning, as @p-j-smith describes. Could comment on this when you have a second. We assume this is what was preventing this pull from being merged. Thanks

bieniekmateusz commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,146 @@
+    timeseries_data = [[] for _ in range(tau_max)]
+
+    # calculate autocorrelation
+    for t in range(0, len(list_of_sets), window_step):
```
I'd love to have a look at the cython part since I never wrote any actual code in it. However, it might be important to ask if this is necessary right now. This code has a linear complexity (window_step is always small) and is significantly faster than the previous implementations. The performance tests I run showed that 99.9% of the running time is spent in atom_select, and the actual Autocorrelation took less than a second. 

So I'd advise against making this code more complex with cython, at least until we have a clear case that the performance optimisation here is useful (so on the ground of Premature Optimisation).

bieniekmateusz replied:
Hi @richardjgowers , I finally had a look at the performance and I designed this little unit test to see whether I should look more into cython. It took below 1 second to compute 100 k frames with the same atoms (worst case), and around 8 seconds for 1 million frames. With this number of frames, it is going to be atom selection that takes most of the time. For this reason I would recommend against cythonising this function for now. 

Also, it appears that having a list of sets as an argument in the function complicates cythonizing.

```
import time
def test_autocorrelation_perfTest():
    # generate 1 million input, worst case scenario (50 same atoms found each time)
    input = [{x for x in range(50)} for x in range(1000 * 1000)]

    st = time.time()
    autocorrelation(input, tau_max=3)
    print('Comuptation Time', time.time() - st)
```

bieniekmateusz commented:
Regarding the API. We realise it is a sensitive class and that you care about the users and we are more than happy to try to account for that. At the same time, by ensuring that only one class takes care neatly of finding Hydrogen Bond, we remove some of the confusion. So overall we hope to simplify the interface and the underlying code, which in the long run should bring more users. 

For example, right now someone might use hbonds.HydrogenBondAnalysis to find hydrogen bonds with a certain set of parameters, just to realise that they cannot calculate HBondLifetime - because it has to find its own Hydrogen Bonds from scratch, without letting the user define them. 

In other words, the temporary pain for the users to switch will hopefully be minimal with the new HBAnalysis in #2237 and allow directly to calculate Lifetime from their dataset.

bieniekmateusz commented:
The separation of the more intuitive Autocorrelation and Intermittency in now separate functions has the potential to provide a good basis for consistency, as the functions are shared across Hydrogen Bond Lifetimes and Survival Probability. 

This is particularly the case with Intermittency which we found to be a tricky concept. By ensuring that atoms are only counted if they comeback, we do not overestimate the presence of atoms. Also, by using the idea of consecutive absence, we ensure that whether it is tau=8 or tau=20, the intermittency of (ie) 2 has exactly the same meaning and significance.

bieniekmateusz commented:
Hi @richardjgowers  

Is there anything we can do to facilitate the discussion or progress here? 

Thanks, Mat

richardjgowers commented:
@bieniekmateusz we can't deprecate `hbondautocorrel` into `hbondanalysis.autocorrel` until the API for defining/finding hydrogen bonds is identical between the two, so that can't merge currently.

I (or someone else) needs to review the autocorrelation code here and PR #2237 , they can probably get merged soon

orbeckst commented:
Given that PR #2237 was merged, what's the status on this PR?

p-j-smith commented:
We should have some time to wrap this up over the next week or so. We'll rebase onto the latest develop and then add a `hydrogenbonds.lifetime` method (which will use our standalone autocorrelation function in this PR) to the new `HydrogenBonds` class.

p-j-smith commented:
Hi @orbeckst it looks like this build failed due to a test for encore.dres_convergence(), which contains random numbers. Is it possible to re-run the build without making another commit and then pushing?

IAlibay commented:
@p-j-smith you can restart Travis by closing and re-opening the PR (just tested on #2563).

p-j-smith commented:
Thanks @IAlibay !

bieniekmateusz commented:
Hi @richardjgowers, would you mind letting us know if there is anything else we should attend to? We'd love to merge this so that we can use it for the hydrogen bond analysis which is just being merged now (#2572)

bieniekmateusz commented:
Hi @orbeckst. It seems that @richardjgowers is quite busy and it has been now quite a few months. Should we consider reassigning this to someone else? 

Thanks, Mat

orbeckst commented:
@bieniekmateusz I think the last couple of weeks were quite hectic with GSoC, so thank you for your patience. I haven't had a close look at the PR to have a real opinion but at the surface it looks as if this is far along.

@richardjgowers – what's your realistic availability over the next week to have a look at this PR again? "No time" is totally acceptable, then we'll try to find someone else.

orbeckst commented:
In the meantime, can you resolve the conflicts for merge?

orbeckst commented:
@bieniekmateusz I haven't reallyyet  reviewed the PR but I noted that the coverage in waterdynamics dropped a lot  by > 10% (and hbonds/hbond_dyamics a little by ~1%). Are these losses real? If so, is there code that needs to be retired/legacy-tested?

bieniekmateusz commented:
Hey @orbeckst , I just had a look at the coverage report with @p-j-smith. Apparently, the moving of the autocorrelation function from waterdynamics.py into utils, which is now tested separately in utils, is the reason for the drop in coverage. 

In hbonds.hbond_analysis.py we added a function that can make use of our autocorrelation function to calculate the hydrogen bond lifetimes. There, the only lines that are not tested are the logging lines and that leads to the 1% decrease in coverage.

orbeckst commented on package/MDAnalysis/analysis/hbonds/hbond_analysis.py:
```diff
@@ -335,6 +335,10 @@ class HydrogenBondAnalysis_OtherFF(HydrogenBondAnalysis):
 
+from ..utils.autocorrelation import autocorrelation, correct_intermittency
+
+from ..utils.autocorrelation import autocorrelation, correct_intermittency
```
duplicated line?

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+
+
+def autocorrelation(list_of_sets, tau_max, window_step=1):
+    r"""The descrete implementation of the autocorrelation function.
```
```suggestion
    r"""The discrete implementation of the autocorrelation function.
```

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+    Parameters
+    ----------
+    list_of_sets : list
+      List of sets,
```
```suggestion
      List of sets
```

orbeckst replied:
What is this?

bieniekmateusz replied:
The list of sets  is a general input that allows to use the same autocorrelation function in both cases, the analysis of atom survival probability, as well as hydrogen bond survival. In the case of atom survival, it is a set that contains IDs of the atoms. In the case of hydrogen bond, the set should actually be a tuple of two atom IDs.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+    r"""The descrete implementation of the autocorrelation function.
+
+    .. math::
+       C_{HB}^c(\tau) = \frac{\sum_{ij}h_{ij}(t_0)h'_{ij}(t_0+\tau)}{\sum_{ij}h_{ij}(t_0)}
```
Can you add some explanation on what the function requires as input? For instance, what does the "list of sets" represent and how does it relate to the mathematical expression?

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+    list_of_sets : list
+      List of sets,
+    tau_max : int
+      The last tau (inclusive) for which to carry out autocorrelation.
```
Name `tau`, e.g., the "lag time"?

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+    """
+
+    # Check types
+    assert type(list_of_sets) == list, "list_of_sets must be a one-dimensional list of sets"
```
Please raise exceptions instead of using `assert` for input checking. This looks like `TypeError`.

(Test for exceptions are very welcome but if you really don't have time, mark them explicitly as not covered ... there's a pragma for coverage to tag blocks or lines of code as explicitly not tested. At least then we know where to send people for more GSoC issues...)

`assert` can be optimized away and it makes for terrible user experience.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+
+    # Check types
+    assert type(list_of_sets) == list, "list_of_sets must be a one-dimensional list of sets"
+    assert type(list_of_sets[0]) == set,  "list_of_sets must be a one-dimensional list of sets"
```
`TypeError`

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+    assert type(list_of_sets[0]) == set,  "list_of_sets must be a one-dimensional list of sets"
+
+    # Check dimensions of parameters
+    assert len(list_of_sets) > tau_max, "tau_max cannot be greater than the length of list_of_sets"
```
`ValueError`

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+
+def correct_intermittency(list_of_sets, intermittency):
+    """
+    Process consecutive intermittency with a single pass over the data. The returned data can be used as input to
```
Doc strings should have a single summary line, empty line, then description.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -0,0 +1,155 @@
+    """
+    Process consecutive intermittency with a single pass over the data. The returned data can be used as input to
+    the function `autocorrelation` in order to calculate the discrete autocorrelation function with a given
+    intermittency.
```
Can you add more text to explain what the function does? Equations, references, etc please.

orbeckst replied:
(I realize that you're not the original author of the function but it will now show up in a user-facing module so we need to be a bit clearer what to do with it.)

bieniekmateusz replied:
Hi @orbeckst , with regards to the intermittency, together with @p-j-smith we are the original authors. This feature was not available for SurvivalProbability before. We will provide more information and hopefully we'll soon write it up as well. Thanks

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -449,12 +445,15 @@
+import warnings
 import numpy as np
 import MDAnalysis.analysis.hbonds
+from MDAnalysis.lib.log import ProgressMeter
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

Please remove.

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -507,7 +506,11 @@ class HydrogenBondLifetimes(object):
+
+    def __init__(self, universe, selection1, selection2, t0, tf, dtmax,
+                 nproc=1):
+        #warnings.warn("This class is deprecated; use analysis.hydrogen_bonds.hbond_analysis.HydrogenBondAnalysis ",
```
Is the class not deprecated anymore? If so, remove the comments.

bieniekmateusz replied:
We aim to depricate it after this PR is merged, due to several problems with the current implementation. Removed the comment for now.

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -808,9 +811,11 @@ def _sameMolecTandDT(self, selection, t0d, tf):
         selection = []
-        for ts in ProgressBar(universe.trajectory, verbose=True,
-                              total=universe.trajectory.n_frames):
+        pm = ProgressMeter(universe.trajectory.n_frames,
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

You seem to be reverting PR #2617!

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -808,9 +811,11 @@ def _sameMolecTandDT(self, selection, t0d, tf):
+                           interval=10, verbose=True)
+        for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
+            pm.echo(ts.frame)
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

You seem to be reverting PR #2617!

bieniekmateusz replied:
Sorry! A sloppy merge on my side. Corrected.

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -981,9 +986,11 @@ def run(self, **kwargs):
         selection = []
-        for ts in ProgressBar(universe.trajectory, verbose=True,
-                              total=universe.trajectory.n_frames):
+        pm = ProgressMeter(universe.trajectory.n_frames,
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

You seem to be reverting PR #2617!

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -981,9 +986,11 @@ def run(self, **kwargs):
+                           interval=10, verbose=True)
+        for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
+            pm.echo(ts.frame)
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

You seem to be reverting PR #2617!

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1121,9 +1128,11 @@ def _sameMolecTandDT(self, selection, t0d, tf):
-        for ts in ProgressBar(universe.trajectory, verbose=True,
-                              total=universe.trajectory.n_frames):
+        pm = ProgressMeter(universe.trajectory.n_frames,
+                           interval=10, verbose=True)
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

You seem to be reverting PR #2617!

orbeckst commented on package/MDAnalysis/analysis/waterdynamics.py:
```diff
@@ -1121,9 +1128,11 @@ def _sameMolecTandDT(self, selection, t0d, tf):
+                           interval=10, verbose=True)
+        for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
+            pm.echo(ts.frame)
```
ProgressMeter should not be here. It's been deprecated and removed everywhere as part of PR #2617 .

You seem to be reverting PR #2617!

bieniekmateusz commented:
Thanks @orbeckst for taking the time. We believe we have answered all of your questions now with the latest push. Please let us know if you have any other comments.

bieniekmateusz commented:
We've added the module under 4.12 Utils, inside of the analysis modules. Please let us know if you are happy with the location or if we should move it somewhere more suitable. I think we now addressed all of your questions. Thanks

bieniekmateusz commented:
After further thought, we think that MDAnalysis.lib.mdamath might be a more suitable location for these functions, but we'll defer to you on this.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
+acount for intermittency before passing the results to
+:func:`autocorrelation`.
+
+See Gowers and Carbonne, 2015, (DOI:10.1063.1.4922445) for a further
```
Please cite using reST syntax and put the paper in references.rst (see [References](https://www.mdanalysis.org/mdanalysis/documentation_pages/references.html).

If the whole module is based on the paper, please also add a  *duecredit* citation to the module or the appropriate function (look elsewhere in the code base where `duecredit` is used – you will only need the DOI for that).

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
+discussion on the time continuous and intermittent autocorrelation of
+hydrogen bond lifetimes.
+
+**Analysis tools that make use of modules**
```
```suggestion
.. rubric:: Analysis tools that make use of modules
```

orbeckst replied:
Alternatively, put it in a `.. SeeAlso::` box?

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
+    be encoded with indicator variables, :math:`0` and :math:`1`, to represent the binary
+    state of said property. This special case is often referred to as the
+    survival probability (:math:`S(\tau)`). As an example, in calculating the survival
+    probability of water molecules within :math:`5 \unicode{x212B}` of a protein, each water
```
```suggestion
    probability of water molecules within 5 Å of a protein, each water
```

orbeckst replied:
The first line of the comment header should contain the UTF-8 magic.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
+    this feature is present at every frame from :math:`t_0` to :math:`N(t0, t_0 + \tau)`.
+    The angular brackets represent an average over all time origins, :math:`t_0`.
+
+    See Araya-Secchi et al., 2014, (https://doi.org/10.1016/j.bpj.2014.05.037)
```
Cite with reST.

orbeckst replied:
If use of this function should cite a paper then you can use a `duecredit` decorator. I don't know the history of the function, so your call.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
 
 def autocorrelation(list_of_sets, tau_max, window_step=1):
-    r"""The descrete implementation of the autocorrelation function.
+    r"""Implementation of a discrete autocorrelation function.
```
Thank you for the explanation, very nice.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
 from copy import deepcopy
 
 
 def autocorrelation(list_of_sets, tau_max, window_step=1):
```
Perhaps call it `discrete_autocorrelation` or `survival_function` — more specific is better and the doc string makes clear that it is not just an acf.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
     window_step : int, optional
-      The step for the t0 to perform autocorrelation (without the overlap). Default is 1.
+      The step size for t0 to perform autocorrelation. Ideally, window_step will be larger than
+       tau_max to ensure independence of each window for which the calculation is performed.
```
```suggestion
     tau_max to ensure independence of each window for which the calculation is performed.
```

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
+        the raw data from which the autocorrelation is computed, i.e :math:`S(\tau)` at each window.
+        This allows the time dependant evolution of :math:`S(\tau)` to be investigated.
 
     .. versionadded:: 0.19.2
```
Needs a versionchanged for the move.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
-    assert type(list_of_sets) == list, "list_of_sets must be a one-dimensional list of sets"
-    assert type(list_of_sets[0]) == set,  "list_of_sets must be a one-dimensional list of sets"
+    # check types
+    if (type(list_of_sets) != list and len(list_of_sets) != 0) or type(list_of_sets[0]) != set:
```
Good! Thanks!

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -88,13 +164,30 @@ def autocorrelation(list_of_sets, tau_max, window_step=1):
+    The returned data can be used as input to the function :func:`autocorrelation` in order
+    to calculate the survival probability with a given intermittency.
+
+    See Gowers and Carbonne, 2015, (DOI:10.1063.1.4922445) for a description of
```
Please reST-cite.

orbeckst commented on package/MDAnalysis/analysis/utils/autocorrelation.py:
```diff
@@ -21,44 +21,120 @@
 # J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
 #
 
```
How about moving it to a new module `lib/correlations.py` – I don't want to bloat `mdamath` even further and the functions here are still fairly specific though independent of the core MDA stuff: thus `lib` is a good place and a separate module is also appropriate.

orbeckst commented on package/doc/sphinx/source/documentation_pages/analysis/autocorrelation.rst:
```diff
@@ -0,0 +1,9 @@
+.. automodule:: MDAnalysis.analysis.utils.autocorrelation
```
I would move it to `MDAnalysis.lib.correlations`.

orbeckst commented on package/doc/sphinx/source/documentation_pages/analysis_modules.rst:
```diff
@@ -147,3 +147,11 @@ Data
 
    analysis/data
+
+Utils
```
I would move it to `MDAnalysis.lib.correlations`.

orbeckst commented:
@bieniekmateusz looking good. Will you also add `duecredit`?

Ping me when I should review again.

@richardjgowers – if you have a moment to also look over the final PR then that would be much appreciated.

bieniekmateusz commented:
Thanks @orbeckst . Our understanding is that currently when someone makes use of SurvivalProbability or hbond.hbond_analysis, the previous `duecredit` will be triggered.

bieniekmateusz commented:
We have now further corrected the citations and we hope to write a small paper in which we will discuss the generalisations we introduce. 

Thanks @orbeckst and @richardjgowers for your help. Please let us know if you have any further comments.

bieniekmateusz commented:
The sphinx warnings have been now corrected, so that is all on our side now. Thanks

orbeckst commented:
@richardjgowers if you have anything to add here then please do; if I don't hear anything over the next 2 days I'll just dismiss your review to move forward. Thanks!

orbeckst commented:
(Maybe we get the PR merged in under a year... thank you @bieniekmateusz and @p-j-smith for not giving up!)

bieniekmateusz commented:
We have updated the tags and the CHANGELOG file. 

In the next stage we hope to use these functions in places that currently duplicate this functionality. In particular, we will add a new HydrogenBondLifetimes function to HydrogenBondsAnalysis, which will also remove the issue #2247. It would also be nice to add a feature that leverages the LeafletFinder tool, combined with this autocorrelation function, to calculate the rate of cholesterol flip-flop in lipid bilayers.

We are planning to write a short software development paper explaining the intermittency function and the generalisation of the autocorrelation function. Given your experience in this field @richardjgowers , the fact that your HBL code was part of our inspiration for this module, and your input in developing this module, we were wondering whether you would be willing to provide feedback and guidance on the technical aspects of the paper as an author on the publication? We realise you must be super busy at the minute, so no worries if not!

Big thanks to both of you @orbeckst  and @richardjgowers  for your help in getting this PR through!

orbeckst commented:
I restarted one CI task that ran out of time.

@richardjgowers are you happy & approve the changes?

bieniekmateusz commented:
Hi again, is there else you'd like us to have a look at? 

Thanks, Mat

orbeckst commented:
@bieniekmateusz @p-j-smith – apologies, this dropped off my radar. Merging/squashing now...

bieniekmateusz commented:
Thanks!
